core/slice.go

```diff
@@ -170,30 +180,32 @@ func (sl *Slice) Append(header *types.Header, domPendingHeader *types.Header, do
 	// Combine subordinates pending header with local pending header
 	pendingHeaderWithTermini := sl.computePendingHeader(types.PendingHeader{Header: localPendingHeader, Termini: newTermini}, domPendingHeader, domOrigin)
 
-	// Call my sub to append the block
-	var newPendingEtxs []types.Transactions
+	// Call my sub to append the block, and collect the rolled up ETXs from that sub
+	localPendingEtxs := []types.Transactions{types.Transactions{}, types.Transactions{}, types.Transactions{}}
+	subPendingEtxs := []types.Transactions{types.Transactions{}, types.Transactions{}, types.Transactions{}}
 	if nodeCtx != common.ZONE_CTX {
-		newPendingEtxs, err = sl.subClients[location.SubIndex()].Append(context.Background(), block.Header(), pendingHeaderWithTermini.Header, domTerminus, td, true, reorg)
+		subPendingEtxs, err = sl.subClients[location.SubIndex()].Append(context.Background(), block.Header(), pendingHeaderWithTermini.Header, domTerminus, td, true, reorg, newInboundEtxs)
 		if err != nil {
 			return nil, err
 		}
-	} else {
-		// If we are a zone, initialize newPendingEtxs
-		newPendingEtxs = []types.Transactions{types.Transactions{}, types.Transactions{}, types.Transactions{}}
+
+		// Cache the subordinate's pending ETXs
+		sl.AddPendingEtxs(types.PendingEtxs{block.Header(), subPendingEtxs})
 	}
 
-	// Add our new ETXs to the newPendingEtxs collection
-	// If this is a coincident block, we need to add the full rollup of ETXs.
-	// Otherwise just send the new ETXs emitted in this block.
-	if isCoincident {
-		etxRollup, err := sl.hc.CollectEtxRollup(block)
-		if err != nil {
-			return nil, fmt.Errorf("unable to get ETX rollup")
-		}
-		newPendingEtxs[nodeCtx] = etxRollup
-	} else {
-		newPendingEtxs[nodeCtx] = block.ExtTransactions()
+	// Combine sub's pending ETXs, sub rollup, and our local ETXs into localPendingEtxs
+	// e.g. localPendingEtxs[ctx]:
+	// * for 'ctx' is dom: empty
+	// * for 'ctx' is local: ETXs emitted in this block
+	// * for 'ctx' is direct sub: replace sub pending ETXs with sub rollup ETXs
+	// * for 'ctx' is indirect sub: copy sub pending ETXs (sub's sub has already been rolled up)
+	for ctx := nodeCtx + 2; ctx < common.HierarchyDepth; ctx++ {
```

Comment:
Basically, there are three steps:
1) I just need to copy the rollup set from everything below my sub (e.g. `nodeCtx + 2` and beyond)
1) I need to manually compute the rollup of my immediate sub (e.g. `nodeCtx + 1`)
1) I need to assign my local pending ETXs (e.g. `nodeCtx`)

Today, step 1 is just the case if I am a prime node, I need to copy the zone rollup. I wrote it this way, so that it might expand to an interlink structure when we get around to that.

I'm open to suggestions if you want it written differently

Reply:
Maybe just supplement the comments.

Reply:
Will do